src/filters/filter_shift.hpp

```diff
@@ -104,30 +116,44 @@ namespace dmxfish::filters {
         virtual void get_output_channels(channel_mapping& map, const std::string& name) override {
             if constexpr (std::is_same<T, uint8_t>::value) {
                 for(size_t i = 0; i < values.size(); i++) {
-                    map.eight_bit_channels[name + ":output_" + std::to_string(i+1)] = &values.at(i);
+		    const auto c_name = name + ":output_" + std::to_string(i+1);
+                    map.eight_bit_channels[c_name] = &values.at(i);
                 }
             } else if constexpr (std::is_same<T, uint16_t>::value) {
                 for(size_t i = 0; i < values.size(); i++) {
-                    map.sixteen_bit_channels[name + ":output_" + std::to_string(i+1)] = &values.at(i);
+		    const auto c_name = name + ":output_" + std::to_string(i+1);
+                    map.sixteen_bit_channels[c_name] = &values.at(i);
                 }
             } else if constexpr (std::is_same<T, double>::value) {
                 for(size_t i = 0; i < values.size(); i++) {
-                    map.float_channels[name + ":output_" + std::to_string(i+1)] = &values.at(i);
+                    const auto c_name = name + ":output_" + std::to_string(i+1);
+		    map.float_channels[c_name] = &values.at(i);
                 }
             } else {
                 for(size_t i = 0; i < values.size(); i++) {
-                    map.color_channels[name + ":output_" + std::to_string(i+1)] = &values.at(i);
+		    const auto c_name = name + ":output_" + std::to_string(i+1);
+                    map.color_channels[c_name] = &values.at(i);
                 }
             }
         }
 
         virtual void update() override {
-            if (last_update + *switch_time <= *time){
-                for(size_t i = values.size() - 1; i > 0; i--){
-                    values.at(i) = values.at(i-1);
+            if (last_update + *switch_time <= *time) {
+                for (size_t i = values.size() - 1; i > 0; i--) {
+                    values.at(i) = values.at(i - 1);
                 }
                 values.at(0) = *input;
-                last_update = last_update + *switch_time;
+                //last_update = last_update + *switch_time;
+                if (*switch_time > 0){
+                    last_update = last_update + std::floor((*time - last_update) / *switch_time) * *switch_time;
+                }
+                // alternative instead of the above
+//                if (*time < last_update + 2 * *switch_time){
+//                    last_update = last_update + *switch_time;
+//                }
```

Comment:
just the basic idea: if switch_time was overshooted (at least twice), then set the actual time as reference